[STEP-2] 콘서트 예약 시스템 시나리오 분석 및 계획 설계

Goddohi opened the pull request:
## 참고 자료

## PR 설명
- api 명세서  : 805ed40
- erd  작성: f0b4984
- 인프라구조도 작성 : f12c6cb

## 리뷰 포인트
1. 현재 좌석 홀드 기준을 예약 테이블으로 가지고 가려고 합니다. 각좌석 별로 홀드및 예약현황을 따로 관리해야하는지 아님 예약으로 조인을 이용하여 이대로 가져가도 될지 궁금합니다.

2. JWT관련되서 도입을 열어두고 있습니다. 대기열을 이용할 경우 jwt와 대기열 토큰을 이용했을때 관리해야할 포인트에 대해서 궁금합니다 . 예를 들어 둘다 중복적인 유저 데이터가 존재하는 부분 입니다.

3. 결제와 예약에서 동시성과 동일한 중복결제및 적립을 방지하기위해서 멱등성도 고려해야 할 포인트라고 생각하나 어느 부분을 고려해서 진행을 해야할지 주요 개념 및 노하우 처럼 조언을 구하고 싶습니다.

4. 대기열이 순차적으로 만약 들어가게만 고려한다면 많은 사용자가 순차적으로 들어가나 안에있는 사용자는 계속 좌석 새로고침과 같은 서비스를 이용하려고 하여 사용량이 늘어나는 경우를 고려해야하는지 와 고려한다면 어떤 점으로 해결해야하는지 주요 개념을 여쭤뵈고 싶습니다.

5. 현재 springboot 컨테이너 한개로 설계를 했으나 nginx등을 이용하여 무중단 배포와 같은 점을 고려해보려고 합니다. 아직 학습을 해보지 않아 제외를 하였지만 무중단 배포에 대한 고려점과 관련 견해를 구하고 싶습니다.

## Definition of Done (DoD)
- [x] 콘서트 예약 시스템 erd 작성 
- [x] 콘서트 예약 시스템 인프라 구성도 작성
- [x] API 명세서 작성
- [ ]  Api 명세서 자동화 진행 (swagger)

Goddohi committed:
doc : api 명세서 추가
- 추후 스웨거로 관리

Goddohi committed:
doc : erd 추가

Goddohi committed:
doc : infra구조도 추가

Isaac482763 commented on docs/API_SPCE.md:
````diff
@@ -0,0 +1,262 @@
+```
+
+**Response**
+```
+{
+ "concertId": "12342",
+ "scheduleId": "14232",
+ "totalSeats": 50,
+ "availableCount": 34,
+ "seats": [
+  { "seatNo": 1 },
+  { "seatNo": 2 },
+  { "seatNo": 5 }
+ ]
+}
````
좌석의 가격이 얼마인지의 정보가 있어야할것같습니다.
각 좌석마다 등급(S,A,B) 정보를 추가하고 각 등급별 가격 정보를 저장하는 필드를 추가하면 좋을것같습니다.

Isaac482763 commented on docs/API_SPCE.md:
````diff
@@ -0,0 +1,262 @@
+### 6. 결제 API
+- 결제 처리하고 결제 내역을 생성하는 API 를 작성합니다.
+- 결제가 완료되면 해당 좌석의 소유권을 유저에게 배정하고 대기열 토큰을 만료시킵니다.
+
+```
+POST /api/payments
+Queue-Token: <issued_token>
+Content-Type: application/json
+{
+  "userId": "uuid",
+  "amount": 150000,
+  "methods": [
+    { "type": "POINT", "amount": 50000 },
+    { "type": "CARD", "amount": 100000, "provider": "HYUNDAI_CARD" }
+  ]
+}
````
결제 API의 request에 가격 정보가 있는데요.
API 호출시에 150000이라는 금액을 어떻게 알수있을까요?
결제 대기중인 예약 목록을 조회할수 있는 API가 있으면 금액을 알수있을것 같은데요.

그리고 결제API는 모든 예약에 대해서 결제를 하는것인지, 특정 reservation_id를 입력받아 해당 예약에 대해서 결제하는것인지도 고민해보시면 좋을것같습니다.

Isaac482763 commented on docs/API_SPCE.md:
````diff
@@ -0,0 +1,262 @@
+```
+
+```
+201 Created
+{
+  "reservation_id": "123124",
+  "concertId": "10421",
+  "scheduleId": "12378",
+  "seats": [
+    { "seatNo": 12 },
+    { "seatNo": 13 },
+    { "seatNo": 14 }
+  ],
+  "heldUntil": "2025-09-07T19:05:00+09:00"
+}
````
예약 완료후에 사용자에게 안내하기 위해서는 금액 필드도 있으면 좋을것같습니다.

Isaac482763 commented:
### 1번 질문.
현재 좌석 홀드 기준을 예약 테이블으로 가지고 가려고 합니다. 각좌석 별로 홀드및 예약현황을 따로 관리해야하는지 아님 예약으로 조인을 이용하여 이대로 가져가도 될지 궁금합니다.
-> 동일한 정보를 양측에서 관리하게되면 결국 동기화를 시켜줘야 해서 오히려 복잡해지거나 데이터 일관성에 문제가 생길수 있습니다.
지금 방식으로 하되 성능을 위해서 인덱싱을 추가하는 방법으로 하시는게 좋을것같습니다.

### 2번 질문.
JWT관련되서 도입을 열어두고 있습니다. 대기열을 이용할 경우 jwt와 대기열 토큰을 이용했을때 관리해야할 포인트에 대해서 궁금합니다 . 예를 들어 둘다 중복적인 유저 데이터가 존재하는 부분 입니다.
-> 좋은 고민입니다.
인증은 JWT에서 온전히 담당합니다.
그러므로 대기열 토큰에는 대기열 상태, 그리고 대기열 토큰이 JWT와 매칭되는지를 확인하기 위한 정보만 있으면 됩니다.
사용자 아이디가 대기열 토큰에 있는것이 불필요하다고 생각된다면 JWT의 sub claim값을 저장해놓을수도 있습니다.
서버에서는 대기열 토큰의 sub값이 JWT의 sub값과 일치하는지로 토큰의 유효성을 판단하면 됩니다.
또한 JWT와, 대기열 토큰 둘 다 만료 시점이 있는데 JWT가 만료되었을때 대기열 토큰은 어떻게 할것인가? 이런 부분도 고민해보면 좋을것같습니다.

### 3번 질문.
결제와 예약에서 동시성과 동일한 중복결제및 적립을 방지하기위해서 멱등성도 고려해야 할 포인트라고 생각하나 어느 부분을 고려해서 진행을 해야할지 주요 개념 및 노하우 처럼 조언을 구하고 싶습니다.
-> 좋은 질문입니다.
중복 처리를 막기위해서 idempotency key를 활용하죠
idempotency key 를 비교해서 이미 처리되었던 요청의 경우 처음 요청의 응답과 동일한 응답을 해줍니다. (재 요청인경우 API내부 로직을 실행하지않고 저장해놨던 응답을 다시 리턴해줌.)
이 때 보통 Redis를 활용해서 응답 결과를 저장해놓습니다.
중요한건 실패하는 경우에도 재요청시에 동일한 결과를 리턴해줘야합니다.
그래서 클라이언트에서는 요청이 실패한 경우 새로운 idempotency key를 만들어서 재요청을 해야합니다.
멱등성을 지키는데 있어서 성공 케이스보다 실패 케이스 때문에 까다롭고 고려해야 할 부분이 많이 생깁니다.
- API 요청이 들어와서 DB작업이나 결제관련으로 외부 PG사 API를 호출 (성공함)
- 클라이언트에게 응답 보내는 과정에서 네트워크 장애로 클라이언트에서는 실패로 인식됨
- 실제 결과와 응답의 불일치 발생
그러면 클라이언트에서는 새로운 idempotency key로 다시 요청을 하겠죠.
이런 상황이 발생 가능하기 때문에 idempotency key만으로는 멱등성을 보장하기 힘듭니다.
그래서 서버 내부 비즈니스 로직 내부에서도 중복을 방지하기 위한 로직이 있어야 합니다.
예를들어 주문 결제의 경우, 요청에서 들어온 주문id가 이미 처리 되었는지 DB에서 확인.
PG사 API 요청전에 이미 처리된 기록이 있는지 확인후에 진행.

### 4번 질문.
대기열이 순차적으로 만약 들어가게만 고려한다면 많은 사용자가 순차적으로 들어가나 안에있는 사용자는 계속 좌석 새로고침과 같은 서비스를 이용하려고 하여 사용량이 늘어나는 경우를 고려해야하는지 와 고려한다면 어떤 점으로 해결해야하는지 주요 개념을 여쭤뵈고 싶습니다.
-> 네 동시접속자 인원을 제한해서 괜찮겠지 라고 생각했는데 예상보다 더 많은 부하가 서버에 걸릴수도 있습니다.
- 말씀하신것처럼 유저들이 너무 과하게 요청을 보내는 경우. -> rate limiting으로 유저별 요청 제한을 둘수있습니다.
- 그럼에도 계산했던것보다 서버 부하가 클때 -> 동시 접속 인원을 동적으로 조절. 서버가 과부하 되지 않으면서도 서버 자원을 최대한 사용할수있도록 동시 접속 인원을 동적으로 조절하여 서버 부하를 일정하게 유지

### 5번 질문
현재 springboot 컨테이너 한개로 설계를 했으나 nginx등을 이용하여 무중단 배포와 같은 점을 고려해보려고 합니다. 아직 학습을 해보지 않아 제외를 하였지만 무중단 배포에 대한 고려점과 관련 견해를 구하고 싶습니다.
-> 네 적용해보시면 학습에 많은 도움이 될것같습니다.
Rolling 배포 방식을 기준으로 말씀드리자면
동일한 인스턴스를 여러개 띄워놨을때 순차적으로 한개씩 업데이트를 해주는 방식입니다.
이 때 nginx가 각 인스턴스의 Health check를 하며 살아있는 인스턴스에만 요청을 전달하여 자동으로 트래픽 분산을 처리해줍니다.
또한 각 인스턴스를 업데이트 하기위해 종료할때는 현재 진행중인 요청을 마무리 하고 종료하기위해 `Spring Boot Graceful Shutdown`을 활용합니다.
그리고 까다로운 부분이 DB 스키마 변경, 마이그레이션 인데요.
필요한 컬럼추가  -> 앱 업데이트 -> DB마이그레이션 -> 불필요한 컬럼삭제 이러한 순서로 진행해야합니다.
그리고 상태 모니터링, CI/CD, Docker 컨테이너 오케스트레이션 등의 주제로 스터디 해보면 좋을것 같습니다.

Goddohi committed:
API 문서 수정
- ERD 에 공연별 좌석 등급및 가격추가
- 예약, 결제 과정에서 좌석별 등급, 가격 표기
- 결제시도시 예약id를 통해 좌석을 예약하는 것으로 명시
